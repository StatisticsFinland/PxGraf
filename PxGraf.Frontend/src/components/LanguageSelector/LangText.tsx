import { Typography } from '@mui/material';
import React from 'react';
import styled from 'styled-components';

interface ILangTextProps {
    text: string;
    underline: boolean;
}

const UnderlinedTypography = styled(Typography)`
    text-decoration: underline;
`;

export const LangText: React.FC<ILangTextProps> = ({ text, underline }) => {
    return underline ? (<UnderlinedTypography>{text}</UnderlinedTypography>) : (<Typography>{text}</Typography>);
}

export default LangText;