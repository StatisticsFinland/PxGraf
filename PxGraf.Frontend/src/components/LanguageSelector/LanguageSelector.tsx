import { Stack, Button } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { UiLanguageContext } from "contexts/uiLanguageContext";
import { LangText } from './LangText';
import styled from 'styled-components';
import Typography from '@mui/material/Typography';
import InfoBubble from 'components/InfoBubble/InfoBubble';

const SelectorWrapper = styled(Stack)`
    padding: 8px;
`;

const StyledSecondaryText = styled(Typography)`
    margin-bottom: -2px;
    margin-top: -2px;
    text-align: right;
`;

const TitleWrapper = styled.div`
    display: flex;
    align-items: center;
`;

const StyledLangButton = styled(Button)`
    && {
        padding: 6px 12px;
        min-width: 0;
    }
`;

export const LanguageSelector: React.FC = () => {
    const { t } = useTranslation();
    const theme = useTheme();
    const { language, setLanguage, availableUiLanguages } = React.useContext(UiLanguageContext);

    return (
        <SelectorWrapper direction="column">
            <TitleWrapper>
                <StyledSecondaryText variant="body2" sx={{ color: theme.palette.text.secondary }}>
                    {t("general.uiLanguage")}
                </StyledSecondaryText>
                <InfoBubble info={t("infoText.langSelector")} ariaLabel={t("general.uiLanguage")} />
            </TitleWrapper>
            <Stack direction="row" flexWrap='wrap'>
                {availableUiLanguages.map(lang => (
                    <StyledLangButton size="small" href='#' aria-label={`${t('general.uiLanguage')}: ${t('lang.' + lang)}`} key={lang} onClick={() => setLanguage(lang)}>
                        <LangText text={lang.toUpperCase()} underline={language === lang} />
                    </StyledLangButton>
                ))}
            </Stack>
        </SelectorWrapper>
    );
}

export default LanguageSelector;
