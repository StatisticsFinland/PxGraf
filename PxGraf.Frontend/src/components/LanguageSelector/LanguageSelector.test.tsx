import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { LanguageSelector } from './LanguageSelector';
import UiLanguageContext from 'contexts/uiLanguageContext';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const setLanguage = jest.fn();
const language = 'fi';
const setLanguageTab = jest.fn();
const languageTab = 'fi';
const availableUiLanguages = ['fi', 'en', 'sv'];
const uiContentLanguage = 'fi';
const setUiContentLanguage = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <LanguageSelector />
            </UiLanguageContext.Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('should invoke setLanguage if clicked', () => {
        render(
            <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                <LanguageSelector />
            </UiLanguageContext.Provider>
        );
        fireEvent.click(screen.getByText('FI'));
        expect(setLanguage).toBeCalledTimes(1);
        fireEvent.click(screen.getByText('SV'));
        expect(setLanguage).toBeCalledTimes(2);
    });
});