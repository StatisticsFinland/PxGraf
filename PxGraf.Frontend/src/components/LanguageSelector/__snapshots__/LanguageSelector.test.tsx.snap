// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root sc-dIMoHT bLAcLM css-nen11g-MuiStack-root"
  >
    <div
      class="sc-gDpztx gmWdfc"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 sc-iQQCXo dzRZkK css-1wle3ir-MuiTypography-root"
      >
        general.uiLanguage
      </p>
      <button
        aria-label="tooltip.open: general.uiLanguage"
        class="sc-fYsHOw lgeTOX"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-ocq0ss-MuiSvgIcon-root"
          data-testid="InfoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
          />
        </svg>
      </button>
    </div>
    <div
      class="MuiStack-root css-1d0f7f9-MuiStack-root"
    >
      <a
        aria-label="general.uiLanguage: lang.fi"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary sc-kpOvIu irOXVc css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
        href="#"
        tabindex="0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 sc-Qotzb cegUxY css-rizt0-MuiTypography-root"
        >
          FI
        </p>
      </a>
      <a
        aria-label="general.uiLanguage: lang.en"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary sc-kpOvIu irOXVc css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
        href="#"
        tabindex="0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          EN
        </p>
      </a>
      <a
        aria-label="general.uiLanguage: lang.sv"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary sc-kpOvIu irOXVc css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
        href="#"
        tabindex="0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
        >
          SV
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
